### PR TITLE
Group Authorization

### LINKED ISSUE #394: Authorization implemented - group section
Description of roles in perun: https://is.muni.cz/auth/th/t4sdm/Master_Thesis_final.pdf
Where to find policies what needs to be used: perun-roles.yml in perun repository(backend)
What needs to be done:
In entity-overview hide items that user dont have role to see them - what does that means - eg. - vo - application item - user shouldnt have see that item if he doesnt have the authority to call the method getApplication(example)
Same rules are applicated to side-menu.
Also check the entity description(there could be link into the user detail or facility detail) and you must check if that link should be available.

Then when we go to the eg. vo-groups page we agreed on:
1. button - create/add - this should be easy look on vo-groups button add, I already did an example how to do that
2. button - remove/delete - when checking/unchecking the select checkbox in the table always check if the user have the authority to delete this group and if dont make the delete button disabled and make the tooltip on that button say that he doesnt have the priviledge to delete some of the selected groups.
3. clicking on the detail of the entity - always check on first entity in the list if the user have priviledge to go into the detail of entity. There is example in vo-groups.

Your fake account that you have created use for the testing if your solution works. If that wouldnt work there would be PerunPriviledge exception notification showing up.

### PR DESCRIPTION
**Pull request by xpavlic**
Resolves #394
Authorization for group section added.

### CODE REVIEW

**Review by xkureck — CHANGES_REQUESTED** [review]
I guess attributes was made by Adam so when you rebase the attributes should work as it should?
image
Because right now I cant see any tooltip why I can select the attributes
image
As a vo observer I can see the checkboxes - that is really weird
image
vo observer: I should see button create subgroup
image
Shouldnt see that
image
wrong tooltip when I want to click on checkbox and also I can click on create even if I dont have nothing checked, I know that it is not part of authorization but please fix that
image
Also please fix bug when I want to get into the detail of application of the group the sidemenu get crazy and it get me into the vo application, which is wrong, I want to stay in the group applications

File: libs/perun/components/src/lib/groups-list/groups-list.component.html [hunk 1]
```diff
@@ -1,8 +1,11 @@
-<div [class.hide-table]="exporting" [hidden]="groups.length === 0 || dataSource.filteredData.length === 0" class="card mt-2">
+<div [class.hide-table]="exporting" [hidden]="groups.length === 0 || dataSource.filteredData.length === 0"
+     class="card mt-2">
   <div class="card-body">
-    <perun-web-apps-table-options (exportEnd)="exporting = false" (exportStart)="exporting = true" [exporter]="exporter"></perun-web-apps-table-options>
+    <perun-web-apps-table-options (exportEnd)="exporting = false" (exportStart)="exporting = true"
+                                  [exporter]="exporter"></perun-web-apps-table-options>
```
Comment (xkureck) [inline libs/perun/components/src/lib/groups-list/groups-list.component.html:5]: one "thing" one row please

File: libs/perun/components/src/lib/groups-list/groups-list.component.html [hunk 2]
```diff
@@ -1,8 +1,11 @@
...
     <div class="overflow-auto">
-      <table #exporter="matTableExporter" [dataSource]="dataSource" class="w-100" mat-table matSort matSortActive="id" matSortDirection="asc" matSortDisableClear
+      <table #exporter="matTableExporter" [dataSource]="dataSource" class="w-100" mat-table matSort matSortActive="id"
+             matSortDirection="asc" matSortDisableClear
```
Comment (xkureck) [inline libs/perun/components/src/lib/groups-list/groups-list.component.html:8]: one "thing" one row please

File: libs/perun/components/src/lib/groups-list/groups-list.component.html [hunk 3]
```diff
@@ -15,15 +18,18 @@
...
-                          [aria-label]="checkboxLabel(row)"
-                          [checked]="selection.isSelected(row)"
-                          [disabled]="(row.name === 'members' && disableMembers) || disableSelect(row.id)"
-                          matTooltip="{{'SHARED_LIB.PERUN.COMPONENTS.GROUPS_LIST.DISABLED_HINT' | translate}}"
-                          [matTooltipDisabled]="!disableSelect(row.id)"
-                          color="primary">
-            </mat-checkbox>
+            <span
+              [matTooltip]="(authType === 'create-relation-dialog' ? 'SHARED_LIB.PERUN.COMPONENTS.GROUPS_LIST.CREATE_RELATION_AUTH_TOOLTIP' :
+               'SHARED_LIB.PERUN.COMPONENTS.GROUPS_LIST.ALREADY_MEMBER_TOOLTIP') | translate"
```
Comment (xkureck) [inline libs/perun/components/src/lib/groups-list/groups-list.component.html:23]: better code readability - translate in in the ts file and here you get only variables

**Review by xkureck — CHANGES_REQUESTED** [review]
image
Hide the outline when the routing is disabled

**Review by bodnara — CHANGES_REQUESTED** [review]
Editing of email footer not working for any role
image

Same goes for editing of notifications
image

As Vo observer, I cannot open notification detail. I assume that this role should be able to see content of the notification only editing should be disabled.
image

After adding member the dialog is not closing. Perhaps you could fix it even if it is not part of authorization.
image

As group-admin after creating subgroup I need to refresh page to gain authorization rights over this group.
Before refresh:
image
After refresh:
image

As group admin I don't have rights to add/remove resources. I am not sure if this is expected behavior, maybe I am missing some information.
image

**Comment by xpavlic** [comment]
image
this is expected behaviour, because you also need RESOURCESELFSERVICE role.
image

File: apps/admin-gui/src/app/shared/components/dialogs/add-edit-notification-dialog/add-edit-notification-dialog.component.ts [hunk 4]
```diff
@@ -45,10 +45,16 @@ export class AddEditNotificationDialogComponent implements OnInit {
...
       };
 
       this.editAuth = this.authResolver.isAuthorized('vo-addMail_ApplicationForm_ApplicationMail_policy', [vo]);
+    } else if (this.data.groupId) {
+      const group = {
+        id: this.data.groupId,
+        beanName: "Group"
+      };
+      this.editAuth = this.authResolver.isAuthorized('group-addMail_ApplicationForm_ApplicationMail_policy', [group]);
     }
```
Comment (xkureck) [inline apps/admin-gui/src/app/shared/components/dialogs/add-edit-notification-dialog/add-edit-notification-dialog.component.ts:58]:
```suggestion
      if (this.data.groupId) {
      this.groupsService.getGroupById(this.data.groupId).subscribe(group => {
        this.editAuth = this.authResolver.isAuthorized('group-addMail_ApplicationForm_ApplicationMail_policy', [group]);
        console.log('edit auth');
        console.log(this.editAuth);
      });
    } else if (this.data.voId) {
      const vo = {
        id: this.data.voId,
        beanName: "Vo"
      };
      this.editAuth = this.authResolver.isAuthorized('vo-addMail_ApplicationForm_ApplicationMail_policy', [vo]);
    }
```
This is suggestion for whole block starting from row 45. The method isAuthorized in this case need whole object group. It needs voId in the group object

**Review by xkureck — CHANGES_REQUESTED** [review]

**Review by xkureck — APPROVED** [review]

**Pull request merged**